File: solutions/java/1074.java
class Solution {
  public int numSubmatrixSumTarget(int[][] matrix, int target) {
    final int m = matrix.length;
    final int n = matrix[0].length;

    int ans = 0;

    for (int[] row : matrix)
      for (int j = 1; j < row.length; ++j)
        row[j] += row[j - 1];

    for (int col = 0; col < n; ++col)
      for (int j = col; j < n; ++j) {
        Map<Integer, Integer> count = new HashMap<>();
        count.put(0, 1);
        int sum = 0;
        for (int i = 0; i < m; ++i) {
          sum += matrix[i][j] - (col > 0 ? matrix[i][col - 1] : 0);
          ans += count.getOrDefault(sum - target, 0);
          count.put(sum, count.getOrDefault(sum, 0) + 1);
        }
      }

    return ans;
  }
}

File: solutions/cpp/1074.cpp
class Solution {
 public:
  int numSubmatrixSumTarget(vector<vector<int>>& matrix, int target) {
    const int m = matrix.size();
    const int n = matrix[0].size();

    int ans = 0;

    for (vector<int>& row : matrix)
      for (int j = 1; j < row.size(); ++j) row[j] += row[j - 1];

    for (int col = 0; col < n; ++col)
      for (int j = col; j < n; ++j) {
        unordered_map<int, int> count;
        count[0] = 1;
        int sum = 0;
        for (int i = 0; i < m; ++i) {
          sum += matrix[i][j] - (col > 0 ? matrix[i][col - 1] : 0);
          ans += count[sum - target];
          ++count[sum];
        }
      }

    return ans;
  }
};